File: example/1d-linear-convection/weno3/cpp/01d/src/cfd_solver.cpp
// ==================== src/cfd_solver.cpp (修复版) ====================
#include "cfd_solver.hpp"

namespace cfd {

    void CfdSolver::initialize(const CfdConfig& config) {
        config_ = config;
        config_.validate();

        // Initialize domain
        domain_ = ComputationalDomain(config_);

        // Initialize solution
        solution_ = Solution(domain_);

        // Create reconstructor
        reconstructor_ = ReconstructorFactory::create_reconstructor(config_);

        // Create flux calculator
        flux_calculator_ = FluxFactory::create_flux_calculator(config_.flux_type);

        // Create boundary condition (default: periodic for convection)
        boundary_condition_ = BoundaryFactory::create_boundary(config_);

        // Create initial condition (default: step function for 1D convection)
        std::vector<Real> step_params = {1.0, 2.0, 0.5, 1.0};  // low, high, start, end
        initial_condition_ = InitialConditionFactory::create_initial_condition("step", step_params);

        // Initialize solution with default conditions
        initialize_solution();

        // Adjust time step based on CFL condition
        calculate_dt();

        if (config_.verbose) {
            std::cout << "CFD solver initialized:" << std::endl;
            std::cout << "  Reconstructor: " << reconstructor_->name() << std::endl;
            std::cout << "  Flux scheme: " << flux_calculator_->name() << std::endl;
            std::cout << "  Boundary condition: " << boundary_condition_->name() << std::endl;
            std::cout << "  Initial condition: " << initial_condition_->name() << std::endl;
            std::cout << "  Time step: " << config_.dt << std::endl;
        }
    }

    void CfdSolver::initialize_solution() {
        // Apply initial condition
        initial_condition_->initialize_with_ghosts(solution_.u(), domain_);

        // Apply boundary condition
        boundary_condition_->apply(solution_.u(), domain_);

        // Save as old solution
        solution_.copy_to_old();
    }

    void CfdSolver::calculate_dt() {
        // CFL condition: dt <= CFL * dx / |wave_speed|
        Real dt_cfl = config_.cfl * domain_.mesh().dx() / std::abs(config_.wave_speed);

        if (config_.dt > dt_cfl) {
            if (config_.verbose) {
                std::cout << "Adjusting time step for stability:" << std::endl;
                std::cout << "  Original dt = " << config_.dt << std::endl;
                std::cout << "  CFL dt = " << dt_cfl << std::endl;
            }
            config_.dt = dt_cfl;
            if (config_.verbose) {
                std::cout << "  Using dt = " << config_.dt << std::endl;
            }
        }
    }

    void CfdSolver::set_boundary_condition(std::unique_ptr<BoundaryCondition> bc) {
        boundary_condition_ = std::move(bc);
        // Re-apply boundary conditions after changing
        boundary_condition_->apply(solution_.u(), domain_);
    }

    void CfdSolver::set_boundary_condition(const std::string& bc_type, 
        Real left_val, Real right_val) {
        boundary_condition_ = BoundaryFactory::create_boundary(bc_type, left_val, right_val);
        // Re-apply boundary conditions after changing
        boundary_condition_->apply(solution_.u(), domain_);
    }

    void CfdSolver::set_initial_condition(std::unique_ptr<InitialCondition> ic) {
        initial_condition_ = std::move(ic);
        // Reinitialize solution
        initialize_solution();
    }

    void CfdSolver::set_initial_condition(const std::string& ic_type,
        const std::vector<Real>& params) {
        initial_condition_ = InitialConditionFactory::create_initial_condition(ic_type, params);
        // Reinitialize solution
        initialize_solution();
    }

    void CfdSolver::compute_residual() {
        // Apply boundary conditions using the boundary condition object
        boundary_condition_->update_ghost_cells(solution_.u(), domain_);

        // Reconstruct face values
        reconstructor_->reconstruct(solution_.u(), 
            solution_.q_face_left(),
            solution_.q_face_right(),
            domain_);

        // Compute fluxes
        flux_calculator_->compute_flux_array(
            solution_.q_face_left(),
            solution_.q_face_right(),
            solution_.flux(),
            config_.wave_speed
        );

        // Compute residual (flux divergence)
        Real dx_inv = 1.0 / domain_.mesh().dx();
        for (int i = 0; i < domain_.mesh().ncells(); ++i) {
            solution_.res()[i] = -(solution_.flux()[i+1] - solution_.flux()[i]) * dx_inv;
        }
    }

    void CfdSolver::runge_kutta_1(Real dt) {
        // Compute residual
        compute_residual();

        // Update solution
        for (int i = domain_.ist(); i <= domain_.ied(); ++i) {
            int j = i - domain_.ist();
            solution_.u()[i] = solution_.u()[i] + dt * solution_.res()[j];
        }

        // Apply boundary conditions
        boundary_condition_->update_ghost_cells(solution_.u(), domain_);

        // Save old solution
        solution_.copy_to_old();
    }

    void CfdSolver::runge_kutta_2(Real dt) {
        // Stage 1
        compute_residual();

        for (int i = domain_.ist(); i <= domain_.ied(); ++i) {
            int j = i - domain_.ist();
            solution_.u()[i] = solution_.u()[i] + dt * solution_.res()[j];
        }
        boundary_condition_->update_ghost_cells(solution_.u(), domain_);

        // Stage 2
        compute_residual();
        for (int i = domain_.ist(); i <= domain_.ied(); ++i) {
            int j = i - domain_.ist();
            solution_.u()[i] = 0.5 * solution_.un()[i] + 
                0.5 * solution_.u()[i] + 
                0.5 * dt * solution_.res()[j];
        }
        boundary_condition_->update_ghost_cells(solution_.u(), domain_);

        solution_.copy_to_old();
    }

    void CfdSolver::advance_one_step(Real dt) {
        switch (config_.rk_order) {
        case 1:
            runge_kutta_1(dt);
            break;
        case 2:
            runge_kutta_2(dt);
            break;
        default:
            throw std::runtime_error("Unsupported RK order: " + std::to_string(config_.rk_order));
        }
    }

    Vector CfdSolver::run_simulation(Real final_time) {
        // Use config final time if not specified
        if (final_time < 0.0) {
            final_time = config_.final_time;
        }

        // Ensure solution is initialized (safety check)
        if (!boundary_condition_ || !initial_condition_) {
            throw std::runtime_error("Solver not properly initialized. Call initialize() first.");
        }

        Real t = 0.0;
        Real dt_old = config_.dt;
        Real dt = dt_old;
        const int max_steps = 100000;  // Safety limit

        if (config_.verbose) {
            std::cout << "Starting time integration..." << std::endl;
            std::cout << "  Final time: " << final_time << std::endl;
            std::cout << "  Time step: " << dt << std::endl;
            std::cout << "  CFL number: " << config_.cfl << std::endl;
            std::cout << "  Reconstructor: " << reconstructor_->name() << std::endl;
            std::cout << "  Flux scheme: " << flux_calculator_->name() << std::endl;
            std::cout << "  Boundary condition: " << boundary_condition_->name() << std::endl;
            std::cout << "  Initial condition: " << initial_condition_->name() << std::endl;
        }

        int step = 0;
        while (t < final_time - 1.0e-12 && step < max_steps) {
            ++step;

            if (t + dt > final_time) {
                dt = final_time - t;
            }

            advance_one_step(dt);
            t += dt;

            // Progress report
            if (config_.verbose && step % config_.print_interval == 0) {
                std::cout << "  Step " << step << ", Time = " << t << std::endl;
            }
        }

        if (step >= max_steps) {
            std::cout << "Warning: Reached maximum number of steps (" << max_steps << ")" << std::endl;
        }

        config_.dt = dt_old;

        if (config_.verbose) {
            std::cout << "Time integration completed:" << std::endl;
            std::cout << "  Total steps: " << step << std::endl;
            std::cout << "  Final time: " << t << std::endl;
        }

        // Return interior solution (without ghost cells)
        return solution_.get_interior_solution();
    }

    Real CfdSolver::analytical_solution(Real x, Real t) const {
        Real L = domain_.mesh().L();
        Real a = config_.wave_speed;

        // Periodic boundary handling
        Real x_shifted = x - a * t;
        while (x_shifted < 0.0) x_shifted += L;
        while (x_shifted > L) x_shifted -= L;

        // Use the current initial condition to compute analytical solution
        return initial_condition_->evaluate(x_shifted);
    }

    Vector CfdSolver::compute_analytical_solution(Real t) const {
        Vector analytical(domain_.mesh().ncells());
        const auto& xcc = domain_.mesh().cell_centers();

        for (int i = 0; i < domain_.mesh().ncells(); ++i) {
            analytical[i] = analytical_solution(xcc[i], t);
        }
        return analytical;
    }

    void CfdSolver::write_solution(const std::string& filename, const Vector& solution) const {
        solution_.write_to_file(filename, solution, "CFD Solution");
    }

    void CfdSolver::write_current_solution(const std::string& filename) const {
        solution_.write_current_solution(filename, "CFD Current Solution");
    }

    void CfdSolver::print_info() const {
        std::cout << "\nCFD Solver Information:" << std::endl;
        std::cout << "==========================================" << std::endl;
        config_.print_brief();
        std::cout << "\n";
        domain_.print_info();
        std::cout << "  Reconstructor: " << reconstructor_->name() << std::endl;
        std::cout << "  Flux scheme: " << flux_calculator_->name() << std::endl;
        std::cout << "  Boundary condition: " << boundary_condition_->name() << std::endl;
        std::cout << "  Initial condition: " << initial_condition_->name() << std::endl;
        std::cout << "  Time integration: RK" << config_.rk_order << std::endl;
        std::cout << "==========================================" << std::endl;
    }

    // Static methods
    Vector CfdSolver::run_single_simulation(const CfdConfig& config) {
        CfdSolver solver(config);
        // 不再需要调用 initialize_step_function()
        return solver.run_simulation();
    }

    void CfdSolver::perform_eno_weno_analysis() {
        std::cout << "==========================================" << std::endl;
        std::cout << "OneFLOW-CFD Solver: ENO3 vs WENO3 Analysis" << std::endl;
        std::cout << "==========================================" << std::endl;

        // Create configurations
        auto [config_eno, config_weno] = CfdConfig::create_analysis_configs();
        config_eno.output_dir = ".";
        config_weno.output_dir = ".";

        // Run ENO simulation
        std::cout << "\nRunning ENO simulation..." << std::endl;
        config_eno.verbose = true;
        Vector u_eno = run_single_simulation(config_eno);

        // Run WENO simulation
        std::cout << "\nRunning WENO simulation..." << std::endl;
        config_weno.verbose = true;
        Vector u_weno = run_single_simulation(config_weno);

        // Create solver for analytical solution
        CfdSolver solver(config_weno);
        Vector u_analytical = solver.compute_analytical_solution(config_weno.final_time);

        // Write results
        std::cout << "\nWriting results to files..." << std::endl;

        solver.write_solution("eno_results.txt", u_eno);
        solver.write_solution("weno_results.txt", u_weno);
        solver.write_solution("analytical_results.txt", u_analytical);

        // Print statistics
        auto eno_minmax = std::minmax_element(u_eno.begin(), u_eno.end());
        auto weno_minmax = std::minmax_element(u_weno.begin(), u_weno.end());
        auto anal_minmax = std::minmax_element(u_analytical.begin(), u_analytical.end());

        std::cout << "\nSimulation statistics:" << std::endl;
        std::cout << "  ENO min/max: " << *eno_minmax.first << " / " << *eno_minmax.second << std::endl;
        std::cout << "  WENO min/max: " << *weno_minmax.first << " / " << *weno_minmax.second << std::endl;
        std::cout << "  Analytical min/max: " << *anal_minmax.first << " / " << *anal_minmax.second << std::endl;

        std::cout << "\nResults written to:" << std::endl;
        std::cout << "  eno_results.txt" << std::endl;
        std::cout << "  weno_results.txt" << std::endl;
        std::cout << "  analytical_results.txt" << std::endl;

        std::cout << "\nTo generate the comparison plot, run:" << std::endl;
        std::cout << "  python scripts/postprocess.py" << std::endl;
        std::cout << "==========================================" << std::endl;
    }

} // namespace cfd

File: example/1d-linear-convection/weno3/fortran/cfd/02s/cfd_solver.f90
! ==================== cfd_solver.f90 ====================
! OneFLOW-CFD Solver for 1D Convection Equation

module cfd_solver
    use, intrinsic :: iso_fortran_env, only: dp => real64
    use config_module, only: CfdConfigType, create_eno_config, create_weno_config
    use mesh_module, only: MeshType
    use domain_module, only: ComputationalDomainType
    use solution_module, only: SolutionType
    use flux_module, only: inviscid_flux
    use boundary_module, only: boundary
    use initial_condition_module, only: initial_condition, analytical_solution
	use recon_module, only: init_coef, ReconType,  &
        eno_reconstruct, weno_reconstruct
    implicit none
    
    private

    ! ===================================================================
    ! Main CFD Solver Class
    ! ===================================================================
    ! ======================================================
    ! 【修改3】：CfdType 中去掉多态 reconstructor，改为两个独立变量
    ! ======================================================
    type, public :: CfdType
        type(CfdConfigType) :: config
        type(ComputationalDomainType) :: domain
        type(SolutionType) :: solution
        type(ReconType), allocatable :: my_recon
    contains
        procedure :: init => cfd_init
    end type CfdType
	
    
    ! ===================================================================
    ! Public Procedures
    ! ===================================================================
    public :: run_simulation, init_field, performEnoWenoAnalysis
    
    
    contains
    
    ! ===================================================================
    ! Initialization Methods
    ! ===================================================================
    
    ! CFD solver initialization
   
    subroutine cfd_init(this, config, domain)
        class(CfdType), intent(inout) :: this
        type(CfdConfigType), intent(in) :: config
        type(ComputationalDomainType), intent(in) :: domain
        
        this%config = config
        this%domain = domain
        call this%solution%init(domain)
		
        allocate(this%my_recon)
        call this%my_recon%init(config, domain)
        
        ! Adjust time step based on CFL condition
        call calculate_dt(this)
    end subroutine cfd_init    
    
    ! Calculate time step based on CFL condition
    subroutine calculate_dt(cfd)
        type(CfdType), intent(inout) :: cfd
        
        real(dp) :: dt_cfl
        
        ! CFL condition: dt <= CFL * dx / |wave_speed|
        dt_cfl = cfd%config%cfl * cfd%domain%mesh%dx / abs(cfd%config%wave_speed)
        
        if (cfd%config%dt > dt_cfl) then
            print *, "Adjusting time step for stability:"
            print *, "  Original dt = ", cfd%config%dt
            print *, "  CFL dt = ", dt_cfl
            cfd%config%dt = dt_cfl
            print *, "  Using dt = ", cfd%config%dt
        end if

    end subroutine calculate_dt
    
    ! ===================================================================
    ! Field Initialization (使用外部模块的函数)
    ! ===================================================================
    
    ! Initialize field with step function
    subroutine init_field(cfd)
        type(CfdType), intent(inout) :: cfd
        integer :: i, j
        
        do i = 1, cfd%domain%mesh%ncells
            ! 使用 initial_condition_module 中的函数
            cfd%solution%u(cfd%domain%ist + i - 1) = initial_condition(cfd%domain%mesh%xcc(i))
        end do
        
        ! 使用 boundary_module 中的函数
        call boundary(cfd%solution%u, cfd%domain%nghosts, cfd%domain%ist, cfd%domain%ied)
        call update_oldfield(cfd%solution%un, cfd%solution%u, cfd%domain%ntcells)
    end subroutine init_field
    
    ! ===================================================================
    ! Reconstruction Methods
    ! ===================================================================
    
    subroutine reconstruction(q, recon, config, domain, solution)
        real(dp), intent(in) :: q(:)
		type(CfdConfigType), intent(in) :: config
        type(ComputationalDomainType), intent(in) :: domain
        type(SolutionType), intent(inout) :: solution
        type(ReconType), intent(inout) :: recon
		
		call recon%reconstruct(q, config, domain, solution)
    end subroutine reconstruction
    
    ! ===================================================================
    ! Residual Computation (更新：使用新的通量模块)
    ! ===================================================================
    
    ! Compute residual (flux divergence)
    subroutine residual(q, cfd)
        real(dp), intent(in) :: q(:)
        type(CfdType), intent(inout) :: cfd
        
        integer :: i
        
        ! Reconstruction using current solution
        call reconstruction(q, cfd%my_recon, cfd%config, cfd%domain, cfd%solution)
        
        ! Compute fluxes - 使用新的通量模块
        call inviscid_flux(cfd%solution%q_face_left, cfd%solution%q_face_right, &
                          cfd%solution%flux, cfd%config, cfd%domain%mesh)
        
        ! Compute residual
        do i = 1, cfd%domain%mesh%ncells
            cfd%solution%res(i) = -(cfd%solution%flux(i+1) - cfd%solution%flux(i)) / &
                                   cfd%domain%mesh%dx
        end do
    end subroutine residual
    
    ! ===================================================================
    ! Time Integration (保持原有实现)
    ! ===================================================================
    
    ! Update old field
    subroutine update_oldfield(qn, q, n)
        real(dp), intent(out) :: qn(:)
        real(dp), intent(in) :: q(:)
        integer, intent(in) :: n
        
        qn(1:n) = q(1:n)
    end subroutine update_oldfield
    
    ! 1st-order Runge-Kutta (Euler)
    subroutine runge_kutta_1(cfd)
        type(CfdType), intent(inout) :: cfd
        
        integer :: i, j
        real(dp) :: dt
        
        dt = cfd%config%dt
        
        ! Compute residual
        call residual(cfd%solution%u, cfd)
        
        ! Update solution
        do i = cfd%domain%ist, cfd%domain%ied - 1
            j = i - cfd%domain%ist + 1
            cfd%solution%u(i) = cfd%solution%u(i) + dt * cfd%solution%res(j)
        end do
        
        ! Apply boundary conditions again
        call boundary(cfd%solution%u, cfd%domain%nghosts, cfd%domain%ist, cfd%domain%ied)
        
        ! Save old solution
        call update_oldfield(cfd%solution%un, cfd%solution%u, cfd%domain%ntcells)
    end subroutine runge_kutta_1
    
    ! 2nd-order Runge-Kutta (Heun)
    subroutine runge_kutta_2(cfd)
        type(CfdType), intent(inout) :: cfd
        
        integer :: i, j
        real(dp) :: dt
        
        dt = cfd%config%dt
        
        ! Stage 1
        call residual(cfd%solution%u, cfd)
        
        do i = cfd%domain%ist, cfd%domain%ied - 1
            j = i - cfd%domain%ist + 1
            cfd%solution%u(i) = cfd%solution%u(i) + dt * cfd%solution%res(j)
        end do
        call boundary(cfd%solution%u, cfd%domain%nghosts, cfd%domain%ist, cfd%domain%ied)
        
        ! Stage 2
        call residual(cfd%solution%u, cfd)
        do i = cfd%domain%ist, cfd%domain%ied - 1
            j = i - cfd%domain%ist + 1
            cfd%solution%u(i) = 0.5_dp * cfd%solution%un(i) + &
                                0.5_dp * cfd%solution%u(i) + &
                                0.5_dp * dt * cfd%solution%res(j)
        end do
        call boundary(cfd%solution%u, cfd%domain%nghosts, cfd%domain%ist, cfd%domain%ied)
        
        call update_oldfield(cfd%solution%un, cfd%solution%u, cfd%domain%ntcells)
    end subroutine runge_kutta_2
    
    ! Runge-Kutta selection
    subroutine runge_kutta(cfd)
        type(CfdType), intent(inout) :: cfd
        
        select case(cfd%config%rk_order)
        case(1)
            call runge_kutta_1(cfd)
        case(2)
            call runge_kutta_2(cfd)
        case default
            call runge_kutta_1(cfd)
        end select
    end subroutine runge_kutta
    
    ! ===================================================================
    ! Simulation Driver
    ! ===================================================================
    
    ! Run simulation to final time
    function run_simulation(cfd, final_time) result(u_result)
        type(CfdType), intent(inout) :: cfd
        real(dp), intent(in) :: final_time
        real(dp), allocatable :: u_result(:)
        
        real(dp) :: t, dt, dt_old
        integer :: step, max_steps
        
        allocate(u_result(cfd%domain%mesh%ncells))
        
        t = 0.0_dp
        dt_old = cfd%config%dt
        dt = dt_old
        max_steps = 10000  ! Safety limit
        
        print *, "Starting time integration..."
        print *, "  Final time: ", final_time
        print *, "  Time step: ", dt
        print *, "  CFL number: ", cfd%config%cfl
        
        step = 0
        do while (t < final_time - 1.0e-12_dp .and. step < max_steps)
            step = step + 1
            
            if (t + dt > final_time) then
                dt = final_time - t
            end if
            
            cfd%config%dt = dt
            call runge_kutta(cfd)
            t = t + dt
            
            ! Progress report
            if (mod(step, 100) == 0) then
                print *, "  Step ", step, ", Time = ", t
            end if
        end do
        
        if (step >= max_steps) then
            print *, "Warning: Reached maximum number of steps (", max_steps, ")"
        end if
        
        cfd%config%dt = dt_old
        
        print *, "Time integration completed:"
        print *, "  Total steps: ", step
        print *, "  Final time: ", t
        
        ! Extract physical solution (without ghost cells)
        u_result = cfd%solution%u(cfd%domain%ist:cfd%domain%ied)
    end function run_simulation
    
    ! ===================================================================
    ! Main Analysis Function
    ! ===================================================================
    
    ! Perform ENO-WENO comparative analysis
    subroutine performEnoWenoAnalysis()
        type(CfdConfigType) :: config_eno3, config_weno3
        type(MeshType) :: mesh
        type(ComputationalDomainType) :: domain_eno3, domain_weno3
        type(CfdType) :: cfd_eno3, cfd_weno3
        real(dp), allocatable :: u_eno(:), u_weno(:), u_analytical(:)
        real(dp), allocatable :: xcc(:)
        integer :: i, ncells, iunit
        
        ! Initialize mesh
        call mesh%init()
        ncells = mesh%ncells
        allocate(xcc(ncells))
        xcc = mesh%xcc
        
        print *, "=========================================="
        print *, "Mesh parameters:"
        print *, "  ncells = ", ncells
        print *, "  dx = ", mesh%dx
        print *, "  L = ", mesh%L
        print *, "=========================================="
        
        ! Configure ENO3
        config_eno3 = create_eno_config()
        
        ! Configure WENO3
        config_weno3 = create_weno_config()
        
        ! Create domains
        call domain_eno3%init(mesh, config_eno3)
        call domain_weno3%init(mesh, config_weno3)
        
        ! Create CFD solvers
        call cfd_eno3%init(config_eno3, domain_eno3)
        call cfd_weno3%init(config_weno3, domain_weno3)
        
        ! Allocate arrays
        allocate(u_eno(ncells), u_weno(ncells), u_analytical(ncells))
        
        ! Run ENO simulation
        print *, "=========================================="
        print *, "Running ENO simulation..."
        print *, "  Scheme: ENO", config_eno3%spatial_order
        print *, "  Time step: ", config_eno3%dt
        print *, "=========================================="
        
        call init_field(cfd_eno3)
        u_eno = run_simulation(cfd_eno3, config_eno3%final_time)
        
        ! Run WENO simulation
        print *, "=========================================="
        print *, "Running WENO simulation..."
        print *, "  Scheme: WENO", config_weno3%spatial_order
        print *, "  Time step: ", config_weno3%dt
        print *, "=========================================="
        
        call init_field(cfd_weno3)
        u_weno = run_simulation(cfd_weno3, config_weno3%final_time)
        
        ! Compute analytical solution
        print *, "Computing analytical solution..."
        do i = 1, ncells
            u_analytical(i) = analytical_solution(xcc(i), config_weno3%final_time, &
                                                config_weno3%wave_speed, mesh%L)
        end do
        
        ! Write results to files
        print *, "Writing results to files..."
        
        ! Write ENO results
        open(newunit=iunit, file='eno_results.txt', status='replace')
        write(iunit, '(A)') '# x, u (ENO)'
        do i = 1, ncells
            write(iunit, '(2F12.6)') xcc(i), u_eno(i)
        end do
        close(iunit)
        
        ! Write WENO results
        open(newunit=iunit, file='weno_results.txt', status='replace')
        write(iunit, '(A)') '# x, u (WENO)'
        do i = 1, ncells
            write(iunit, '(2F12.6)') xcc(i), u_weno(i)
        end do
        close(iunit)
        
        ! Write analytical results
        open(newunit=iunit, file='analytical_results.txt', status='replace')
        write(iunit, '(A)') '# x, u (Analytical)'
        do i = 1, ncells
            write(iunit, '(2F12.6)') xcc(i), u_analytical(i)
        end do
        close(iunit)
        
        ! Print some statistics
        print *, "=========================================="
        print *, "Simulation statistics:"
        print *, "  ENO min/max: ", minval(u_eno), maxval(u_eno)
        print *, "  WENO min/max: ", minval(u_weno), maxval(u_weno)
        print *, "  Analytical min/max: ", minval(u_analytical), maxval(u_analytical)
        print *, "=========================================="
        
        print *, "Simulation completed successfully!"
        print *, "Results written to:"
        print *, "  eno_results.txt"
        print *, "  weno_results.txt"
        print *, "  analytical_results.txt"
        print *, ""
        print *, "To generate the comparison plot, run:"
        print *, "  python postprocess.py"
        print *, "=========================================="
        
        deallocate(u_eno, u_weno, u_analytical, xcc)
    end subroutine performEnoWenoAnalysis

end module cfd_solver

! ===================================================================
! Main Program
! ===================================================================
program main
    use cfd_solver
    implicit none
    
    print *, "=========================================="
    print *, "OneFLOW-CFD Solver for 1D Convection"
    print *, "ENO vs WENO Comparison"
    print *, "=========================================="
    
    call performEnoWenoAnalysis()
    
    print *, "Program finished successfully!"
    
end program main